File: apps/chaja-app/src/ipc/branches.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

export type BranchKind = "local" | "remote";

export interface BranchInfo {
  name: string;
  full_name: string;
  kind: BranchKind;
  tip_sha: string;
  is_head: boolean;
  upstream: string | null;
  ahead: number;
  behind: number;
}

export function listBranches(repoPath: string): Promise<BranchInfo[]> {
  return invoke<BranchInfo[]>("list_branches", { repoPath });
}

export function createBranch(
  repoPath: string,
  name: string,
  from?: string,
): Promise<void> {
  return invoke<void>("create_branch", { repoPath, name, from });
}

export function deleteLocalBranch(
  repoPath: string,
  name: string,
  force = false,
): Promise<void> {
  return invoke<void>("delete_local_branch", { repoPath, name, force });
}

export function renameBranch(
  repoPath: string,
  oldName: string,
  newName: string,
): Promise<void> {
  return invoke<void>("rename_branch", { repoPath, oldName, newName });
}

/// Rebase the current branch onto `targetBranch`. Aborts on conflict
/// (chajá doesn't yet expose per-step resolution).
export function rebaseCurrentOnto(
  repoPath: string,
  targetBranch: string,
): Promise<void> {
  return invoke<void>("rebase_current_onto", { repoPath, targetBranch });
}

/// Set or clear the upstream tracking config for a local branch.
/// Pass `null` to clear, or `"origin/main"` to track that ref.
export function setUpstream(
  repoPath: string,
  branchName: string,
  upstream: string | null,
): Promise<void> {
  return invoke<void>("set_upstream", { repoPath, branchName, upstream });
}

File: crates/chaja-bridge/src/commands/branches.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

use std::path::PathBuf;

use crate::backend::{BranchInfo, GitBackend};
use crate::repo::GixBackend;

#[tauri::command]
pub async fn list_branches(repo_path: String) -> Result<Vec<BranchInfo>, String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .list_branches(&PathBuf::from(&repo_path))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn create_branch(
    repo_path: String,
    name: String,
    from: Option<String>,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .create_branch(&PathBuf::from(&repo_path), &name, from.as_deref())
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn delete_local_branch(
    repo_path: String,
    name: String,
    force: Option<bool>,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .delete_local_branch(&PathBuf::from(&repo_path), &name, force.unwrap_or(false))
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

#[tauri::command]
pub async fn rename_branch(
    repo_path: String,
    old_name: String,
    new_name: String,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        GixBackend
            .rename_branch(&PathBuf::from(&repo_path), &old_name, &new_name)
            .map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}
